shared/engine/session.js

```diff
@@ -0,0 +1,157 @@
+import type {invokeType} from './platform-specific'
+import {IncomingRequest, OutgoingRequest} from './request'
+import {ConstantsStatusCode} from '../constants/types/flow-types'
+import {rpcLog} from './platform-specific'
+
+// A session is a series of calls back and forth tied together with a single sessionID
+class Session {
+  // Our id
+  _id: SessionID;
+  // Map of methods => callbacks
+  _incomingCallMap: incomingCallMapType;
+  // Let the outside know we're waiting
+  _waitingHandler: ?WaitingHandlerType;
+  // Tell engine we're done
+  _endHandler: ?EndHandlerType;
+  // Sequence IDs we've seen. Value is true if we've responded (often we get cancel after we've replied)
+  _seqIDResponded: {[key: string]: boolean} = {};
+  // If you want to know about being cancelled
+  _cancelHandler: ?CancelHandlerType;
+  // If true this session exists forever
+  _dangling: boolean;
+  // Name of the start method, just to help debug
+  _startMethod: ?MethodKey;
+  // Start callback so we can cancel our own callback
+  _startCallback: ?(err: RPCError, ...args: Array<any>) => void;
+
+  // Allow us to make calls
+  _invoke: invokeType;
+
+  // Outstanding requests
+  _outgoingRequests: Array<Object> = [];
+  _incomingRequests: Array<Object> = [];
+
+  constructor (
+    sessionID: SessionID,
+    incomingCallMap: ?incomingCallMapType,
+    waitingHandler: ?WaitingHandlerType,
+    invoke: invokeType,
+    endHandler: EndHandlerType,
+    cancelHandler?: ?CancelHandlerType,
+    dangling?: boolean = false,
+  ) {
+    this._id = sessionID
+    this._incomingCallMap = incomingCallMap || {}
+    this._waitingHandler = waitingHandler
+    this._invoke = invoke
+    this._endHandler = endHandler
+    this._cancelHandler = cancelHandler
+    this._dangling = dangling
+  }
+
+  set id (sessionID: SessionID) { throw new Error("Can't set sessionID") }
+  get id (): SessionID { return this._id }
+  get dangling (): boolean { return this._dangling }
+
+  // Make a waiting handler for the request. We add additional data before calling the parent waitingHandler
+  // and do internal bookkeeping if the request is done
+  _makeWaitingHandler (isOutgoing: boolean, method: MethodKey, seqid: ?number) {
+    return (waiting: boolean) => {
+      rpcLog('engineInternal', 'waiting state change', this.id, waiting, method, this, seqid)
+      // Call the outer handler with all the params it needs
+      this._waitingHandler && this._waitingHandler(waiting, method, this._id)
+
+      // Request is finished, do cleanup
+      if (!waiting) {
+        const requests = isOutgoing ? this._outgoingRequests : this._incomingRequests
+        const idx = requests.findIndex(r => r.method === method)
+        if (idx !== -1) {
+          // Mark us as responded
+          if (seqid) {
+            this._seqIDResponded[String(seqid)] = true
+          }
+          // Remove from our list
+          requests.splice(idx, 1)
```

Comment:
actually I see you probably want seqIDResponded only true if you had this in your known requests.   So tweak that demo a bit